uncertainty_mapping/draw_uncertainty.py

```diff
@@ -0,0 +1,196 @@
+
+    reader.open(
+        rosbag2_py.StorageOptions(uri=bag_path, storage_id="mcap"),
+        rosbag2_py.ConverterOptions(
+            input_serialization_format="cdr", output_serialization_format="cdr"
+        ),
+    )
+    topic_types = reader.get_all_topics_and_types()
+
+    topic_type_dict = {topic.name: topic.type for topic in topic_types}
+
+    if topic_name not in topic_type_dict:
+        print(f"Topic '{topic_name}' not found in the bag.")
+        return []
+
+    print(f"Reading messages from topic '{topic_name}':")
+
+    traces = []
+    msg_type = get_message(topic_type_dict[topic_name])
+    while reader.has_next():
+        (topic, data, timestamp) = reader.read_next()
+        if topic == topic_name:
+            # Assuming the message type is a plain string
+            traces.append(deserialize_message(data, msg_type))
+    return traces
+
+
+def generate_ellipse_points(center, width, height, angle, num_points=100):
+    """Generate points along the perimeter of an ellipse."""
+    cx, cy = center
+    t = np.linspace(0, 2 * np.pi, num_points)
+    ellipse_points = np.array([width * np.cos(t), height * np.sin(t)]).T
+
+    # Rotate points by the given angle
+    R = np.array(
+        [[np.cos(angle), -np.sin(angle)], [np.sin(angle), np.cos(angle)]]
+    )
+    rotated_points = np.dot(ellipse_points, R)
+
+    # Translate points to the center
+    ellipse_points = rotated_points + np.array([cx, cy])
+    return [[pt[1], pt[0]] for pt in ellipse_points]
+
+
+def draw_uncertainty_ellipse(
+    m,
+    center,
+    x_uncertainty,
+    y_uncertainty,
+    angle=0,
+    color="grey",
+    fill_color="grey",
+    fill_opacity=0.2,
+):
+    """Draw an ellipse of uncertainty on a folium map."""
+    # Convert uncertainty from meters to degrees (approximate)
+    lat_per_meter = (
+        1 / 111320
+    )  # One degree of latitude is approximately 111.32 km
+    lon_per_meter = 1 / (
+        40075000 * np.cos(np.radians(center[0])) / 360
+    )  # One degree of longitude varies based on latitude
+
+    width = x_uncertainty * lon_per_meter
+    height = y_uncertainty * lat_per_meter
+
+    # Generate ellipse points
+    ellipse_points = generate_ellipse_points(
+        (center[1], center[0]), width, height, np.radians(angle)
+    )
+
+    # Convert ellipse points to list of tuples
+    ellipse_points_list = list(map(tuple, ellipse_points))
+
+    # Add ellipse to the map as a polygon
+    folium.Polygon(
+        locations=ellipse_points_list,
+        color=color,
+        fill=True,
+        fill_color=fill_color,
+        fill_opacity=fill_opacity,
+    ).add_to(m)
+
+
+# Assumed to be Novatel BESTPOS.msg    https://github.com/novatel/novatel_oem7_driver/blob/master/src/novatel_oem7_msgs/msg/BESTPOS.msg
+def convert_trace(gps_pos):
+    return GpsPoint(
+        lat=gps_pos.lat,
+        lon=gps_pos.lon,
+        x_uncertainty=gps_pos.lon_sigma,
+        y_uncertainty=gps_pos.lat_sigma,
+    )
```

Comment:
From the `/sensor/gps/bestpos` topic, the actual msg type is [novatel_gps_msgs/msg/NovatelPosition](https://github.com/swri-robotics/novatel_gps_driver/blob/ros2-devel/novatel_gps_msgs/msg/NovatelPosition.msg)

I guess we should have specified earlier that we are using the  [novatel_gps_driver](https://github.com/swri-robotics/novatel_gps_driver/tree/ros2-devel) from swri_robitics rather than the origal developed by Novatel ;P

The msg should be similar as the script works anyways.